training_utils.py

```diff
@@ -257,19 +264,39 @@ def train_loop(
 
                 # Tree regularizer!
                 if num_steps % regularizer_steps == 0:
+                    # Linear scheduling for gumbel softmax temperature tau
+                    tau = tau_init + tau_step * num_steps
                     if (regularizer is not None):
                         if (regularize_all):
-                            sci_charts = regularizer.build_scores(accum_strings, model, 0, tqdm_disable=True, parse_splits=None, batch=True)
+                            sci_charts, samples = regularizer.build_scores(accum_strings, model, 0, tqdm_disable=True, parse_splits=None, batch=True, use_gold=use_gold)
                         else:
-                            sci_charts = regularizer.build_scores(curr_batch_dict['string'], model, 0, tqdm_disable=True, parse_splits=None, batch=True)
+                            sci_charts, samples = regularizer.build_scores(curr_batch_dict['string'], model, 0, tqdm_disable=True, parse_splits=None, batch=True, use_gold=use_gold)
                         if (args.mean_regularize):
-                            sci_scores = regularizer.get_score(sci_charts, mean=True)
+                            sci_scores = regularizer.get_score(sci_charts, mean=True, input_str=samples, use_gold=use_gold, tau=tau)
                         else:
-                            sci_scores = regularizer.get_score(sci_charts, mean=False)
+                            sci_scores = regularizer.get_score(sci_charts, mean=False, input_str=samples, use_gold=use_gold, tau=tau)
                         # print(sci_scores)
-                        fin_sci_score = torch.mean(torch.stack(sci_scores))
-                        sci_loss = -regularizer_rel_wt * fin_sci_score
+                        if (use_linear):
+                            # Shikhar's linear time idea
+                            fin_sci_score = torch.mean(torch.stack([torch.abs(_) for _ in sci_scores]))
+                            sci_loss = regularizer_rel_wt * fin_sci_score
+                        else:
+                            fin_sci_score = torch.mean(torch.stack(sci_scores))
+                            sci_loss = -regularizer_rel_wt * fin_sci_score
                         sci_loss.backward()
+                        if compute_dot:
+                            # Dot products between tree reg gradients and task loss gradients
+                            parameters = [
+                                p for p in model.model.parameters() if p.requires_grad
+                            ]
+                            sci_grads = []
+                            for p in parameters:
+                                if (p.grad == None):
+                                    sci_grads.append(None)
+                                else:
+                                    param = p.grad.cpu().data
+                                    sci_grads.append(param)
+                            model.model.zero_grad()
```

Comment:
Is there a method of reducing the code duplication from these lines and `310-326`?

Reply:
Fixed, thanks for pointing out!